cpp/src/arrow/compute/kernels/util_internal.cc

```diff
@@ -57,22 +57,67 @@ PrimitiveArg GetPrimitiveArg(const ArrayData& arr) {
   return arg;
 }
 
-ArrayKernelExec TrivialScalarUnaryAsArraysExec(ArrayKernelExec exec,
-                                               NullHandling::type null_handling) {
-  return [=](KernelContext* ctx, const ExecBatch& batch, Datum* out) -> Status {
-    if (out->is_array()) {
-      return exec(ctx, batch, out);
+// TODO(wesm): ARROW-16577: this will be unneeded later
```

Comment:
This function is a huge mess but after ARROW-16577 it will not be needed at all, so during that patch (eliminating the `ValueDescr::SCALAR` output path and frankly the whole ValueDescr struct altogether) is the right time to clean this up and move the scalar-to-array-of-length-1 promotion into the kernel execution layer